src/reseeding.rs

```diff
@@ -39,73 +40,127 @@ use {RngCore, SeedableRng, Error, ErrorKind};
+        let results_empty = R::Results::default();
+        ReseedingRng(
+            BlockRng {
+                core: Reseeder {
+                    core: rng,
+                    reseeder: reseeder,
+                    threshold: threshold as i64,
+                    bytes_until_reseed: threshold as i64,
+                },
+                index: results_empty.as_ref().len(), // generate on first use
+                results: results_empty,
+            }
+        )
+    }
+
+    /// Reseed the internal PRNG.
+    pub fn reseed(&mut self) -> Result<(), Error> {
+        self.0.core.reseed()
+    }
+}
+
+impl<R: BlockRngCore<u32> + SeedableRng, Rsdr: RngCore> RngCore for ReseedingRng<R, Rsdr> {
+    #[inline(always)]
+    fn next_u32(&mut self) -> u32 {
+        self.0.next_u32()
+    }
+
+    #[inline(always)]
+    fn next_u64(&mut self) -> u64 {
+        self.0.next_u64()
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.0.fill_bytes(dest)
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+        self.0.try_fill_bytes(dest)
+    }
+}
+
+#[derive(Debug)]
+struct Reseeder<R, Rsdr> {
+    core: R,
+    reseeder: Rsdr,
+    threshold: i64,
+    bytes_until_reseed: i64,
+}
+
+impl<R, Rsdr> BlockRngCore<u32> for Reseeder<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
+    type Results = <R as BlockRngCore<u32>>::Results;
+
+    fn generate(&mut self, results: &mut Self::Results) -> Result<(), Error> {
+        if self.bytes_until_reseed <= 0 {
+            // We want to reseed here, and generate results later in the
+            // function. If generating results fail, we should return the error
+            // from that. If generating results succeeded, but reseeding failed,
+            // we should return the error from reseeding.
+            // The only way to get this behaviour without destroying performance
+            // was to split part of the function out into a
+            // `reseed_and_generate` method.
+            return self.reseed_and_generate(results);
         }
+        self.bytes_until_reseed -= results.as_ref().len() as i64 * 4;
+        self.core.generate(results)
     }
+}
 
+impl<R, Rsdr> Reseeder<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
     /// Reseed the internal PRNG.
-    ///
-    /// This will try to work around errors in the RNG used for reseeding
-    /// intelligently through some combination of retrying and delaying
-    /// reseeding until later. So long as the internal PRNG doesn't fail, this
-    /// method will not fail, i.e. failures from the reseeding source are not
-    /// fatal.
-    pub fn reseed(&mut self) {
-        // Behaviour is identical to `try_reseed`; we just squelch the error.
-        let _res = self.try_reseed();
+    fn reseed(&mut self) -> Result<(), Error> {
+        R::from_rng(&mut self.reseeder).map(|result| self.core = result)
     }
```

Comment:
... this is called by `ReseedingRng::reseed` (though it probably shouldn't be) and by `auto_reseed`. Just inline it; we don't need the extra function. Then you can rename `auto_reseed` → `reseed`.

Reply:
O, I forgot about this function in the first comment, and also with the recent work around `ReseedingRng`.

Before, `ReseedingRng` had a method users could manually call to force the RNG to reseed. Now the `reseed` method has gotten much smarter (thanks to you). Delaying, and wrapping the error. That is very useful for the situations where we reseed 'because it is time'. But I think it useful to also expose the basic functionality again.

That is why the extra function in `ReseedingRng::reseed` (and Reseeder::reseed)...

Reply:
All the "smart" logic does is potentially delay automatic reseeding and change the error kind. I don't see how that is a problem here. It still "forces" reseeding just as much as the old logic; it just acts a bit differently if the attempt fails.